첫번째, 두번째(java-calculator) 숙제 제출

yenny07 opened the pull request.

yenny07 committed:
첫번째 숙제

yenny07 committed:
두번째 숙제

yenny07 committed:
두번째 숙제 테스트 코드 작성

yenny07 committed:
리드미 작성

KimGyeong commented on src/main/java/Operation.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public void checkValue(String value) {
+        if (value.matches("\\d*(\\.\\d+)?")) {
```
이 정규표현식은 어떤것을 의미할까요...? 이것을 상수화시키면 어떨까요?

KimGyeong commented on src/main/java/Operation.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public void operation(int operand) {
+        switch (operator) {
+            case "+":
+                add(operand);
+                break;
+            case "-":
+                minus(operand);
+                break;
+            case "*":
+                multiple(operand);
+                break;
+            case "/":
+                divide(operand);
+                break;
```
operator가 연산자가 아닌 다른 값이 들어오면 어떻게 될까요?

KimGyeong commented on src/main/java/Operation.java:
```diff
@@ -0,0 +1,60 @@
+                firstExists = true;
+            }
+        } else {
+            operator = value;
```
번역하면 `연산자 = 값` 이 되는데, 말이 맞지 않네요? 변수명의 의미를 생각해주세요.

KimGyeong commented on src/main/java/Operation.java:
```diff
@@ -0,0 +1,60 @@
+public class Operation {
+    private int result = 0;
+    private String operator;
+    private boolean firstExists = false;
```
boolean 변수의 이름은 일반적으로 `is~`로 이름을 짓습니다. 그러면 타입을 보지 않고도 어떤 의미를 갖는지 이름만으로 알 수 있겠죠?

KimGyeong commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,20 @@
+
+public class StringCalculator {
+
+    static String[] values = new String[0];
```
Operation이라는 객체에게 주는게 어떨까요? static으로 선언하게 된다면, 두번째 입력을 받을 때 어떻게 될까요?

KimGyeong reviewed (commented):
현종과 전혀 다른 코드여서 재밌네요! 여러 가지를 생각하면서 코드를 작성한 것이 보였어요.
코멘트들과 더불어 제약사항을 추가할게요.
1. 입력에 대한 예외 처리를 해주세요. 불가능한 입력이 발생한 경우, 사용자가 어떤 것이 잘못됐는지 알려주세요.
2. 계산을 끝내고 다시 입력을 받을 수 있게 해주세요. 특정 값을 입력한 경우에 프로그램을 종료할 수 있게 해주세요. 

현재 작업한 브랜치에서 작업한 후에 푸시하고 리뷰 요청해주세요. :)

yenny07 committed:
불명확한 변수명, 정규표현 수정

FirstExists -> isFirst
정규표현 -> ^[0-9]*$
checkValue () -> checkValueType()

yenny07 committed:
적절하지 않은 연산자 예외처리

yenny07 committed:
적절하지 않은 피연산지 예외처리, 입력 반복 추가

yenny07 committed:
Auto stash before merge of "yenny07" and "origin/yenny07"

yenny07 committed:
리드미 수정

KimGyeong commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,30 @@
+            if(input.equals("-1")){
+                break;
+            }
+            operation = new Operation();
+            operation.setValues(input);
```
생성자에서 input을 받는다면 코드가 줄어들겠죠?
그렇게 하기 위해선 Operation이 어떻게 변경되어야 할 지 생각해봅시다.

KimGyeong commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,30 @@
+            operation.setValues(input);
+
+            try{
+                operation.checkValidation();
+                operation.runOperation();
```
main 함수가 이 두가지 과정을 모두 알아야 할까요? 
계산 시작! 이라고 명령하면 계산이 되게끔 수정해봅시다.

KimGyeong replied:
아래의 `getResult`를 사용하지 않고, `runOperation`이 결과를 반환하게끔 수정하는 것도 좋을 것 같네요.
operation이 결과를 가지고 있을 필요는 없으니까요.

KimGyeong commented on src/main/java/Operation.java:
```diff
@@ -0,0 +1,77 @@
+
+    public void runOperation() {
+        for (int i = 0; i < values.length; i++) {
+            if (i % 2 == 0) {
+                if (isFirst) {
+                    result = Integer.parseInt(values[i]);
+                    isFirst = false;
+                } else {
+                    operation(Integer.parseInt(values[i]));
+                }
+            } else {
+                operator = values[i];
+            }
```
이를 private메소드로 추출하면 어떨까요? 너무 길어서 읽기가 힘들다는 생각이 드네요.

KimGyeong reviewed (commented):
피드백이 많이 있지는 않네요. 하지만 하나 하나가 생각보다 큰 변화를 요구할거에요.
객체지향 생활체조를 지키면서 코딩을 하게 될텐데, 9번 항목에 `Getter/Setter/Property를 사용하지 않는다`가 있어요. 이를 선행해본다 생각하시고 수정해보시면 좋을 것 같습니다.
수정한 후에 다시 요청해주세요! :)

yenny07 committed:
refactor : 메소드 분리 및 단순화

- getter 삭제
- setValues() -> public Operator(String input)
- runOperation() 단순화
- main 단순화

yenny07 committed:
test : 테스트코드 수정

KimGyeong commented on src/main/java/Operation.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public Operation(String input) {
+        values = new String[0];
+        values = input.split(" ");
```
values 값을 두 번이나 바꾸는 이유는 뭘까요...?

yenny07 replied:
그렇네 객체를 다시 할당했는데 왜 저걸 그대로 뒀지..

KimGyeong commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,36 @@
+        int result = 0;
+
+        try{
+            operation.checkValidation();
```
`checkValidation` 메소드를 StringCalculator가 알아야 할 필요가 있을까요?

KimGyeong commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,36 @@
+            if(input.equals("-1")){
+                break;
+            }
+            operation = new Operation(input);
```
> 자동차는 자동차 이름, 자동차 번호를 가지고 있고, 자동차는 달리고 멈추는 기능이 있다. 여기에서 가지고 있는 것을 속성이라고 한다. 자바에서는 이러한 속성을 필드(Field)라는 용어로 사용한다.

Operation을 생성하면서 input을 통해 계산식을 필드로 갖고 있는 형태인데, '계산기가 계산식을 가지고 있나요?'라는 생각을 한번 해보시면 어떨까 싶어요. 조금 더 원초적으로 돌아가서 계산기가 어떤 것을 갖고 있는지, 어떤 기능을 갖고 있는지 생각해봅시다.

KimGyeong commented on src/test/java/StringCalculatorTest.java:
```diff
@@ -0,0 +1,39 @@
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+
+class StringCalculatorTest {
+    private static Stream<Arguments> provideInputData() {
+        return Stream.of(
+                Arguments.of("2 * 6 - 1 * 2", 22),
+                Arguments.of("13", 13),
+                Arguments.of("10 - 2 * 2", 16)
+        );
+    }
```
👍

KimGyeong commented on src/test/java/StringCalculatorTest.java:
```diff
@@ -0,0 +1,39 @@
+    @Test
+    @DisplayName("입력된 식의 띄어쓰기가 잘못되었을 때 예외처리")
+    public void testCheckValidation() {
+        String inputdata = "12+ 4";
```
```suggestion
        String inputdata = "12 + 4";
```
Auto Formatter를 사용합시다. 에디터의 `ctrl + alt + l`, 혹은 커밋 탭에서의 세팅을 확인해보세요.

KimGyeong reviewed (commented):
저번에 남긴 리뷰의 도착점이 생각했던 것과 달랐네요. 이번에는 코드를 수정하기 전에 두가지에 대한 답을 스스로 가져보시면 좋겠습니다. 우리 팀원들과 얘기를 해봐도 좋구요.
1. 클래스는 무엇일까요?
2. 객체는 무엇일까요?
수정하고 리뷰요청해주세요~

yenny07 committed:
refactor : 빈 메소드 삭제, 생성자 위치 조정

yenny07 committed:
refactor : main의 간결화

checkValidation() 메소드 호출을 Operation클래스로 이동
사람은 식을 입력하고 '=' 버튼만 누른다..

yenny07 committed:
refactor : 메소드들의 기능 분리

yenny07 committed:
test : 테스트 코드 수정

yenny07 committed:
docs : 리드미 수정

KimGyeong commented on src/main/java/Operation.java:
```diff
@@ -0,0 +1,59 @@
+
+        try {
+            checkValidation(values);
+            result = Integer.parseInt(values[0]);
+            result = getResult(values, result, operator);
```
```suggestion
            result = getResult(values, Integer.parseInt(values[0]), operator);
```
코드는 언제나 짧은게 좋습니당

KimGyeong replied:
```suggestion
            result = getResult(values, operator);
```
`getResult` 메소드에서 values를 파라미터로 받으니, 내부에서 0번째 값만 꺼낼 수도 있겠네요?

KimGyeong reviewed (approved):
아주 많이 좋아졌어요! 처음 제출했을 때의 코드와 비교해보는 시간을 가져봤으면 좋겠습니다. 만약 보람을 느끼지 못했다면... 제 잘못이 큰거겠죠...?
하나의 코멘트를 남겼지만 마지막 리뷰였습니다! 체크해주시고 변경하시면 바로 머지할게요~ 고생했어요!

yenny07 committed:
refactor : 최초항도 getResult()에서 계산하도록 수정
